src/sentry/utils/versioning.py

```diff
@@ -0,0 +1,44 @@
+from __future__ import absolute_import
+
+import warnings
+from collections import namedtuple
+
+from sentry.exceptions import InvalidConfiguration
+
+
+class Version(namedtuple('Version', 'major minor patch')):
+    def __str__(self):
+        return '.'.join(map(str, self))
+
+
+def make_upgrade_message(service, modality, version, hosts):
+    return '{service} {modality} be upgraded to {version} on {hosts}.'.format(
+        hosts=', '.join('{0} (currently {1})'.format(*i) for i in hosts),
+        modality=modality,
+        service=service,
+        version=version,
+    )
+
+
+def check_versions(service, versions, required, recommended=None):
+    """
+    Check that all members of mapping hosts fulfill version requirements.
```

Comment:
Oops, this should probably read `Check that hosts fulfill version requirements.`